x/paychan/handler.go

```diff
@@ -0,0 +1,295 @@
+	}
+
+	if msg.Payment.Amount.Compare(*pc.Total) > 0 {
+		return msg, ErrInvalidAmount(msg.Payment.Amount)
+	}
+	// Payment is representing a cumulative amount that is to be
+	// transferred to recipients account. Because it is cumulative, every
+	// transfer request must be greater than the previous one.
+	if msg.Payment.Amount.Compare(*pc.Transferred) <= 0 {
+		return msg, ErrInvalidAmount(msg.Payment.Amount)
+	}
+
+	return msg, nil
+}
+
+func (h *transferPaymentChannelHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	pc, err := h.bucket.GetPaymentChannel(db, msg.Payment.ChannelId)
+	if err != nil {
+		return res, err
+	}
+
+	// Payment amount is total amount that should be transferred from
+	// payment channel to recipient. Deduct already transferred funds and
+	// move only the difference.
+	diff, err := msg.Payment.Amount.Add(pc.Transferred.Negative())
+	if err != nil || diff.IsZero() {
+		return res, ErrInvalidAmount(msg.Payment.Amount)
+	}
+
+	src := paymentChannelAccount(msg.Payment.ChannelId)
+	dst := weave.NewAddress(pc.Recipient)
+	if err := h.cash.MoveCoins(db, src, dst, diff); err != nil {
+		return res, err
+	}
+
+	// Track total amount transferred from the payment channel to the
+	// recipients account.
+	pc.Transferred = msg.Payment.Amount
+
+	// We care about the latest memo only. Full history can be always
+	// rebuild from the blockchain.
+	pc.Memo = msg.Payment.Memo
+
+	// If all funds were transferred, we can close the payment channel
+	// because there is no further use for it. In addition, because all the
+	// funds were used, no party is interested in closing it.
+	//
+	// To avoid "empty" payment channels in our database, delete it without
+	// waiting for the explicit close request.
+	if pc.Transferred.Equals(*pc.Total) {
+		err := h.bucket.Delete(db, msg.Payment.ChannelId)
+		return res, err
+	}
+
+	obj := orm.NewSimpleObj(msg.Payment.ChannelId, pc)
+	err = h.bucket.Save(db, obj)
+	return res, err
+}
+
+type closePaymentChannelHandler struct {
+	auth   x.Authenticator
+	bucket PaymentChannelBucket
+	cash   cash.Controller
+}
+
+var _ weave.Handler = (*closePaymentChannelHandler)(nil)
+
+func (h *closePaymentChannelHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	_, err := h.validate(ctx, db, tx)
+	return res, err
+}
+
+func (h *closePaymentChannelHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	pc, err := h.bucket.GetPaymentChannel(db, msg.ChannelId)
+	if err != nil {
+		return res, err
+	}
+
+	// If payment channel funds were exhausted anyone is free to close it.
+	if pc.Total.Equals(*pc.Transferred) {
```

Comment:
That was my initial plan, but then I decided to require explicit close. The reason for this is that explicit close is easier for everyone. 
Implementation is simpler, because there are less cases. 
Client interaction is simpler, because you don't have to track how much is there left on the payment channel. Otherwise, closing non existing (automatically closed) channel will return you an error and can cause confusion.

On top of all this, I think it's better to have a nice history and an explicit `close` event recorded.

@alpe  Does this make sense?

Reply:
I like explicit close, easier to track when querying the blockchain history (which is just the transactions/events)

Reply:
I will leave it as it is now then. @alpe is that ok?